src/rust/engine/fs/store/src/snapshot.rs

```diff
@@ -202,358 +200,7 @@ impl Snapshot {
     .boxed()
   }
 
-  ///
-  /// Given N Snapshots, returns a new Snapshot that merges them.
-  ///
-  /// Any files that exist in multiple Snapshots will cause this method to fail: the assumption
-  /// behind this behaviour is that almost any colliding file would represent a Rule implementation
-  /// error, and in cases where overwriting a file is desirable, explicitly removing a duplicated
-  /// copy should be straightforward.
-  ///
-  pub async fn merge(store: Store, snapshots: &[Snapshot]) -> Result<Snapshot, String> {
-    // We dedupe PathStats by their symbolic names, as those will be their names within the
-    // `Directory` structure. Only `Dir+Dir` collisions are legal.
-    let path_stats = {
-      let mut uniq_paths: IndexMap<PathBuf, PathStat> = IndexMap::new();
-      for path_stat in Iterator::flatten(snapshots.iter().map(|s| s.path_stats.iter().cloned())) {
```

Comment:
> That improves its performance compared to what's on master, which will traverse all recursive file and directory nodes of all k directories to merge, and not take advantage of the shared merkle tree implicit structure for performance.

This isn't the case, AFAICT. Merge already used the merkle tree structure. `merge_directories` only recurses into a directory if it needs to to resolve conflicts: if there are no conflicts (as determined by the `if dir_digests.len() == 1` check at the top of `merge_directories_recursive`), then it immediately uses the directory without recursing into it.

I can look more deeply at the new implementation of merge, but it seems like there might be a misunderstanding that would be good to clear up first. It would also be good to ensure that you've run the benchmark to compare an apples to apples case between master and the branch.

Reply:
I did run the benchmark before and after, and it reported a speedup. I'll separate the criterion benchmarks into separate commits before the performance improvements so this can be reproduced.

Regardless of anything else, this PR:
1. Moves higher-level snapshot operations into a separate file, reducing bloat in `snapshot.rs`,
2. Avoids spreading the logic for calculating conflicts in two separate methods, with an `IndexMap` in `merge()` and a separate `error_for_collisions()` method at the bottom.
3. Comments the separate sections of the logic so it's clear why and how it terminates.
4. Uses more async methods.

I think that makes it a useful change, and I'd be extremely glad to spend some time refactoring this further if you think this approach is missing something.

Reply:
Sorry, to be clear: after reviewing `merge_directories_recursive()`, your comment seems completely correct. So I'm going to split out the criterion benchmark commits so I can confirm this.

Also, the `subset()` operation is I think probably what the greater speedup here is due to the `/**` and `/**/*` optimization. It's possible that `merge()` is exactly as performant. Thank you for taking the time to elaborate here.

Reply:
So from the criterion benchmark difference between the two commits now, I'm finding (very surprisingly) that the snapshot merge operation is also significantly faster:
```
snapshot_merge/snapshot_merge
                        time:   [9.8825 ms 9.9829 ms 10.053 ms]
                        change: [-99.343% -99.332% -99.319%] (p = 0.00 < 0.05)
                        Performance has improved.
```

However, note that the one change I had to make to the code of the merge benchmark is to validate that the merges succeed, instead of fail. I believe that when duplication checking occurs, the failure path short-circuits, so this difference is probably a fluke.

However, the snapshot subset benchmark improvement should be legit, and that's the one we're going for, so I'm satisfied with that improvement. I will edit the OP to make it clear that the merge operation was not improved since I agree that the "merkle tree structure" was already being used.

Reply:
...wait. That's the wrong direction. If failure short-circuits, then it would be faster -- but we're still seeing a speedup.

I'm not sure why it might be faster, then, but I also haven't validated, for example, that failure *does* short-circuit. As above, I think I would prefer to leave this as-is since it doesn't seem to be *harming* merge performance, and I think there is benefit to the current PR in refactoring the merge code to make it easier to read.

Reply:
> As above, I think I would prefer to leave this as-is since it doesn't seem to be harming merge performance, and I think there is benefit to the current PR in refactoring the merge code to make it easier to read.

I'd prefer that it was broken out of this PR, since the PR is already quite large, and it's not related to snapshot subsetting, which is the thing that we know to be pathological. Leaving the `merge` benchmark might be good to help enable landing further optimizations though.

Reply:
That's a great idea -- I'll do that.